snuba/utils/streams/kafka.py

```diff
@@ -46,34 +77,76 @@ class KafkaConsumer(Consumer[TopicPartition, int, bytes]):
 
     def __init__(self, configuration: Mapping[str, Any]) -> None:
         self.__consumer = ConfluentConsumer(configuration)
-
-    def __wrap_assignment_callback(
-        self, callback: Callable[[Sequence[TopicPartition]], None]
-    ) -> Callable[[ConfluentConsumer, Sequence[ConfluentTopicPartition]], None]:
-        def wrapper(
-            consumer: ConfluentConsumer, partitions: Sequence[ConfluentTopicPartition]
-        ) -> None:
-            callback(
-                [TopicPartition(value.topic, value.partition) for value in partitions]
-            )
-
-        return wrapper
+        self.__auto_offset_reset = get_enum_configuration_value(
+            configuration,
+            "auto.offset.reset",
+            {
+                "earliest": set(["smallest", "beginning"]),
+                "latest": set(["largest", "end"]),
+                "error": set(),
+            },
+            "largest",
+        )
+        self.__offsets: MutableMapping[TopicPartition, int] = {}
+        self.__state: ConsumerState = ConsumerState.CONSUMING
 
     def subscribe(
         self,
         topics: Sequence[str],
         on_assign: Optional[Callable[[Sequence[TopicPartition]], None]] = None,
         on_revoke: Optional[Callable[[Sequence[TopicPartition]], None]] = None,
     ) -> None:
-        kwargs = {}
+        def assignment_callback(
+            consumer: ConfluentConsumer, partitions: Sequence[ConfluentTopicPartition]
+        ) -> None:
+            self.__state = ConsumerState.ASSIGNING
+
+            # TODO: This might be able to use the ``positions`` API to avoid
+            # rolling back offsets after an uncommitted rebalance?
+            offsets: MutableMapping[TopicPartition, int] = {}
+
+            for i in self.__consumer.committed(partitions):
+                stream = TopicPartition(i.topic, i.partition)
+                if i.offset >= 0:
+                    offsets[stream] = i.offset
+                elif i.offset == OFFSET_INVALID:
+                    # TODO: What arguments should be provided to ``get_watermark_offsets``?
+                    earliest, latest = self.__consumer.get_watermark_offsets(i)
+                    if self.__auto_offset_reset == "earliest":
+                        offsets[stream] = earliest
+                    elif self.__auto_offset_reset == "latest":
+                        offsets[stream] = latest
+                    else:
+                        raise ValueError("unexpected value for `auto.offset.reset`")
+                else:
+                    raise ValueError("received unexpected offset")
```

Comment:
This is basically the same as the `TaskSetConsumer` and the `SynchronizedConsumer`, and probably should have a big comment as such, similar to:

https://github.com/getsentry/snuba/blob/aaef1c6c3e648531132743f695a6d14e80baf9e4/snuba/subscriptions/consumer.py#L317-L324

This should also be mentioned in the docstring to make sure that users are informed about the difference in behavior.